### DISCUSSION TITLE
[FR] Integrate Tailwind into CP

### DISCUSSION THREAD

**Discussion by wbrowar** [craftcms/cms#8836] [Ideas]
### Description

Tailwind seems like a good match for developers who use Craft and there are many plugins that extend the CP’s look and feel. While you're exploring refactoring the CP for Craft 4, it would be great if Tailwind (or a similar Utility CSS library) could be used as the base for the CP UI.

This way developers could extend that design system to match the look and feel of the CP, while leaving control of the design system to Craft.

I can see uses for this in my own plugins, and the idea of using a utility-based design system alongside custom element index views could be very awesome.

**Comment by Anubarak** [comment]
Please no utility first css style. I rather like the clean and convenient CSS of the CP. You don't need to understand much and can pretty much copy paste classes with ease.
I don't want to change simple classes from

```html
<div class="field">
    <div class="heading">
    </div>
    <div class=input">
     </div>
</div>
```
to something hellish as a list of 10 classes each and remember what I need to use for what. If you want to build the CSS with tailwind and keep the clean classes, go for it but please don't use the utility first approach

**Comment by wbrowar** [comment]
@Anubarak that's a good point that it would cause issues in trying to override existing styles and would have downsides from things, like specificity, if it were used everywhere. Maybe refactoring all of Craft’s UI to use something like Tailwind isn't a good idea.

I still think there's a use case for UI layouts and elements created for plugins and custom modules. There is a SCSS mixins file that helps in this situation, but it requires a build process to integrate the SCSS code into your plugin or module. For plugins I feel like that's fine, but when creating a custom module, it would be a big time saver to have a flexible design system in place that could be used and would stay up to date as Craft’s UI is updated.

**Comment by brandonkelly** [comment]
We do plan on adding Tailwind, but that doesn’t mean we’ll get rid of our current classes; those can just be redefined as custom utility classes.

**Comment by piotrpog** [comment]
@brandonkelly Do you plan that for Craft 4 or for some earlier updates?

**Comment by brandonkelly** [comment]
Probably 3.6

**Comment by leevigraham** [comment]
Can't wait to see how you handle purgecss and custom modules / plugins 😃

**Comment by brandonkelly** [comment]
I don’t think we’d use purgecss considering we don’t know which classes plugins may need. We could potentially come up with a way for individual projects to do their own asset packaging though, and provide alternative CSS/JS files for the control panel.

**Comment by wbrowar** [comment]
I wonder if the new Tailwind JIT compiler could be a way to solve for keeping the Tailwind asset file size down. I don't know the best way to implant it so that P&T and plugin developers could use it, but here's an example project where the JIT compiler is done on the fly: https://twitter.com/marcelpociot/status/1387513006351953920

**Comment by wbrowar** [comment]
So, at this point I'm kind of glad this hasn't been added to the CP. I think CSS has come a long way over the past 3 years and now I think the need for utility classes isn't as much of an advantage as it felt like back in 2020.. ~~I would like to see more use of CSS Custom Properties throughout the CP in a way that makes it easy for a plugin to pick up, but that feels like a separate FR and that doesn't sound like high priority work until some sort of CP redesign is on the way..~~ (Okay, I should have checked this before posting.. I see that there are lots of good CSS custom properties to work from.)

Thanks for taking the time to consider this.

**Reply by brandonkelly** [reply]
Definitely need to continue pushing for more CSS variables, as it will be pretty important with dark mode support. Whether we go full Tailwind is yet to be determined.

**Comment by leevigraham** [comment]
I'm with @wbrowar (latest comment) I'm glad this isn't in the CP. More html helpers + CSS variables are the way to go.
